src/stream/src/executor/aggregation/approx_count_distinct.rs

```diff
@@ -0,0 +1,282 @@
+            0
+        };
+
+        Self {
+            registers: [RegisterBucket::new(); INDICES],
+            initial_count: count,
+        }
+    }
+
+    /// Adds the count of the datum's hash into the register, if it is greater than the existing
+    /// count at the register.
+    fn update_registers(&mut self, datum_ref: DatumRef, is_insert: bool) {
+        if datum_ref.is_none() {
+            return;
+        }
+
+        let scalar_impl = datum_ref.unwrap().into_scalar_impl();
+        let hash = self.get_hash(scalar_impl);
+
+        let index = (hash as usize) & (INDICES - 1); // Index is based on last few bits
+        let count = self.count_hash(hash) as usize;
+
+        self.registers[index].update_register(count, is_insert);
+    }
+
+    /// Calculate the hash of the `scalar_impl` using Rust's default hasher
+    /// Perhaps a different hash like Murmur2 could be used instead for optimization?
+    fn get_hash(&self, scalar_impl: ScalarImpl) -> u64 {
+        let mut hasher = DefaultHasher::new();
+        scalar_impl.hash(&mut hasher);
+        hasher.finish()
+    }
+
+    /// Counts the number of trailing zeroes plus 1 in the non-index bits of the hash.
+    fn count_hash(&self, mut hash: u64) -> u8 {
+        let mut count = 1;
+
+        hash >>= INDEX_BITS; // Ignore bits used as index for the hash
+        hash |= 1 << COUNT_BITS; // To allow hash to terminate if it is all 0s
+
+        while hash & 1 == 0 {
+            count += 1;
+            hash >>= 1;
+        }
+
+        count
+    }
+}
+
+impl StreamingAggStateImpl for StreamingApproxCountDistinct {
+    fn apply_batch(
+        &mut self,
+        ops: Ops<'_>,
+        visibility: Option<&Bitmap>,
+        data: &[&ArrayImpl],
+    ) -> Result<()> {
+        match visibility {
+            None => {
+                for (op, datum) in ops.iter().zip_eq(data[0].iter()) {
+                    match op {
+                        Op::Insert | Op::UpdateInsert => self.update_registers(datum, true),
+                        Op::Delete | Op::UpdateDelete => self.update_registers(datum, false),
+                    }
+                }
+            }
+            Some(visibility) => {
+                for ((visible, op), datum) in
+                    visibility.iter().zip_eq(ops.iter()).zip_eq(data[0].iter())
+                {
+                    if visible {
+                        match op {
+                            Op::Insert | Op::UpdateInsert => self.update_registers(datum, true),
+                            Op::Delete | Op::UpdateDelete => self.update_registers(datum, false),
+                        }
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn get_output(&self) -> Result<Datum> {
+        // Approximation for bias correction. See "HyperLogLog: the analysis of a near-optimal
+        // cardinality estimation algorithm" by Philippe Flajolet et al.
+        let bias_correction = 0.72134;
```

Comment:
Are you sure this bias correction is chosen for 2^10 registers?

I calculate 0.7213/(1+(1.079/(2^10))) ~ 0.72054 (see page 140 of https://hal.archives-ouvertes.fr/file/index/docid/406166/filename/FlFuGaMe07.pdf)

Reply:
I chose that bias correction because the additional bias is negligible compared to the estimation error (according to page 8 of https://oertl.github.io/hyperloglog-sketch-estimation-paper/paper/paper.pdf). But I agree it wouldn't hurt to make it more accurate to the number of registers used.